ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java

```diff
@@ -0,0 +1,1760 @@
+      assertEquals("Index segment mapped state not as expected",
+          !indexSegmentStartOffset.equals(lastIndexSegmentStartOffset), indexSegmentEntry.getValue().isMapped());
+    }
+    // verify sanity of compacted index
+    state.verifyRealIndexSanity();
+    // verify ordering
+    assertEquals("Ordering after compaction is inconsistent", validLogEntriesInOrder,
+        getLogEntriesInOrder(targetSegmentNames));
+    // no clean shutdown file should exist
+    assertFalse("Clean shutdown file not deleted",
+        new File(tempDirStr, BlobStoreCompactor.TARGET_INDEX_CLEAN_SHUTDOWN_FILE_NAME).exists());
+    // there should be no temp files
+    assertEquals("There are some temp log segments", 0,
+        tempDir.listFiles(BlobStoreCompactor.TEMP_LOG_SEGMENTS_FILTER).length);
+  }
+
+  /**
+   * Verifies data after compaction is finished.
+   * 1. Checks that GET of data of live keys succeeds without any flags.
+   * 2. Checks that GET of data of deleted keys returns deleted. Also verifies that when GET is used with
+   * {@link StoreGetOptions#Store_Include_Deleted}, data is returned if it was not compacted and not returned if
+   * compacted.
+   * 3. Checks that GET of data of expired keys returns expired if not compacted and not found if compacted. Also
+   * verifies that when GET is used with {@link StoreGetOptions#Store_Include_Expired}, data is returned if it was not
+   * compacted and not returned if compacted.
+   * @param idsInCompactedLogSegments the ids in the compacted segments.
+   * @param deleteReferenceTimeMs the reference time in ms to use to decide whether deletes are valid.
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws StoreException
+   */
+  private void verifyDataPostCompaction(Set<MockId> idsInCompactedLogSegments, long deleteReferenceTimeMs)
+      throws InterruptedException, IOException, StoreException {
+    for (MockId id : state.allKeys.keySet()) {
+      if (state.liveKeys.contains(id)) {
+        BlobReadOptions options = state.index.getBlobReadInfo(id, EnumSet.noneOf(StoreGetOptions.class));
+        checkRecord(id, options);
+        options.close();
+      } else if (state.deletedKeys.contains(id)) {
+        boolean shouldBeCompacted =
+            idsInCompactedLogSegments.contains(id) && state.isDeletedAt(id, deleteReferenceTimeMs);
+        try {
+          state.index.getBlobReadInfo(id, EnumSet.noneOf(StoreGetOptions.class));
+          fail("Should not be able to GET " + id);
+        } catch (StoreException e) {
+          assertEquals(id + " failed with error code " + e.getErrorCode(), StoreErrorCodes.ID_Deleted,
+              e.getErrorCode());
+        }
+        try {
+          BlobReadOptions options = state.index.getBlobReadInfo(id, EnumSet.allOf(StoreGetOptions.class));
+          if (shouldBeCompacted) {
+            fail("Should not be able to GET " + id);
+          } else {
+            checkRecord(id, options);
+            options.close();
```

Comment:
Cannot move it because `options` is in the scope of the `try` only (and that is one that throws the `StoreException`). Could declare options outside the `try` but will have to do null checks in the `finally` (and also add `try..finally` for the good case). Considering that doing this is not really required in the first place, I think all the extra code will make the function a bit unreadable. Would prefer to leave it this way if you don't have any strong objections.